### PR TITLE
[1.9] Replace HealingSourceData with EnderDragon.getHealingCrystal().

### PR DESCRIPTION
**Pull request by JBYoshi**
**SpongeAPI** | [SpongeCommon](https://github.com/SpongePowered/SpongeCommon/pull/756)
I heard @gabizou say some time ago that `DataManipulator`s are no longer allowed to store `Entity`s. So instead of implementing `HealingSourceData`, I just moved it into `EnderDragon` and implemented it there.

### CODE REVIEW

File: src/main/java/org/spongepowered/api/entity/living/complex/EnderDragon.java [hunk 1]
```diff
@@ -24,17 +24,25 @@
...
+
 /**
  * Represents an Ender Dragon.
  */
 public interface EnderDragon extends ComplexLiving, Boss, Aerial {
 
     @Override
     Set<EnderDragonPart> getParts();
+    
+    Optional<EnderCrystal> getHealingCrystal();
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/entity/living/complex/EnderDragon.java]: Javadoc would be useful

Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/entity/living/complex/EnderDragon.java]: Whoops, completely forgot. Thanks for reminding me!

Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/entity/living/complex/EnderDragon.java]: Fixed.

File: src/main/java/org/spongepowered/api/entity/living/complex/EnderDragon.java [hunk 2]
```diff
@@ -24,17 +24,33 @@
...
  * Represents an Ender Dragon.
  */
 public interface EnderDragon extends ComplexLiving, Boss, Aerial {
 
     @Override
     Set<EnderDragonPart> getParts();
+    
+    /**
+     * Returns the current {@code EnderCrystal} that is healing this ender dragon.
+     * @return The ender crystal
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/entity/living/complex/EnderDragon.java]:
```java
* Returns the current {@code EnderCrystal} that is healing this ender dragon.
*
* @return The ender crystal
```
(new line)

File: src/main/java/org/spongepowered/api/entity/living/complex/EnderDragon.java [hunk 3]
```diff
@@ -24,17 +24,33 @@
...
+     */
+    Optional<EnderCrystal> getHealingCrystal();
+    
+    /**
+     * Sets the current {@code EnderCrystal} that is healing this ender dragon.
+     * @param crystal The ender crystal
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/entity/living/complex/EnderDragon.java]:
```java
* Sets the current {@code EnderCrystal} that is healing this ender dragon.
*
* @param crystal The ender crystal
```
(new line)

File: src/main/java/org/spongepowered/api/entity/living/complex/EnderDragon.java (hunk 2 above)
Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/entity/living/complex/EnderDragon.java]: Fixed.

File: src/main/java/org/spongepowered/api/entity/living/complex/EnderDragon.java (hunk 3 above)
Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/entity/living/complex/EnderDragon.java]: Fixed.

**Comment by Aaron1011** [comment]
@JBYoshi: For future reference, `DataManipulators` only store persistent data. If they need to store entities, they can store a `UUID` instead of the entity itself.

However, since the healing crystal isn't persistent, this PR is fine.

**Comment by Aaron1011** [comment]
While the API itself looks fine, there's an issue with how the healing crystal works in Vanilla.

The current crystal is recalculated about twice a second (a 1/10 chance every time `onLivingUpdate` is called). This is done by finding the closest crystal to the dragon, and setting that as the current healing crystal. So, any crystal set through the API will be quickly overridden (unless it happens to be the closest to the dragon).

Further complicating matters is the fact that the selected crystal isn't even sent to the client. Instead, the client calculates the nearest crystal on its own, and draws the beam animation between the dragon and that crystal. While the server and client values are allways synchronized in vanilla, they would become synchronized if the plugin assigns a different healing crystal.

@Zidane @gabizou: Thoughts on how we should deal with this?

**Comment by Zidane** [comment]
@Aaron1011 

I agree with your assessment and while its entirely possible to set the crystal, it really does no good for the plugin...it'll just be overwritten and the client says fuck all and chooses its own anyway.

Keep the getter, toss the setter.

**Comment by JBYoshi** [comment]
@Aaron1011 @Zidane Updated.

File: src/main/java/org/spongepowered/api/entity/living/complex/EnderDragon.java [hunk 4]
```diff
@@ -24,17 +24,28 @@
...
 package org.spongepowered.api.entity.living.complex;
 
+import org.spongepowered.api.entity.EnderCrystal;
 import org.spongepowered.api.entity.living.Aerial;
 import org.spongepowered.api.entity.living.monster.Boss;
 
+import java.util.Optional;
 import java.util.Set;
 
+import javax.annotation.Nullable;
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/entity/living/complex/EnderDragon.java]: Unnecessary import.

**Comment by ryantheleach** [comment]
Would the setter be useful for fake dragons used for effect that didn't have an AI? Or would they be considered a different entity.

**Pull request merged**